wallet-new/src/Cardano/Wallet/Kernel/Submission.hs

```diff
@@ -0,0 +1,431 @@
+                toEntry (txId, txAux) = ScheduledTx txId txAux (SubmissionCount 0)
+            in Schedule (IntMap.insertWith mappend (castSlot $ mapSlot succ currentSlot) txs s) n
+
+-- | Purges all the scheduled transactions in the given 'Slot'. This is
+-- automatically called by 'tick', and it's responsibility of @rho@ to
+-- correctly reschedule all the transaction.
+purgeSlot :: Slot -> Schedule -> Schedule
+purgeSlot slot (Schedule s n) = Schedule (IntMap.delete (castSlot slot) s) n
+
+-- Ready-to-use 'ResubmissionFunction's.
+
+-- | A 'RetryPolicy' is simply a function which instruct the 'Schedule' when
+-- to attempt resubmitting the given 'ScheduledTx' item.
+type RetryPolicy = SubmissionCount -> Slot -> Maybe Slot
+
+limited :: MaxRetries -> RetryPolicy -> RetryPolicy
+limited maxRetries retryPolicy = \retries currentSlot ->
+    case getSubmissionCount retries >= maxRetries of
+         True  -> Nothing
+         False -> retryPolicy retries currentSlot
+
+type Exponent   = Double
+type MaxRetries = Int
+
+--
+-- Stock retry policies inspired by the 'retry' package.
+--
+
+-- | Very simple policy which merely retries to resubmit the very next 'Slot',
+-- up until 'MaxRetries' attempts.
+constantRetry :: MaxRetries -> RetryPolicy
+constantRetry maxRetries =
+    limited maxRetries $ \_ currentSlot -> Just (mapSlot succ currentSlot)
+
+-- | An exponential backoff policy, parametric over a maximum number of
+-- 'MaxRetries' and an 'Exponent' for the backoff.
+exponentialBackoff :: MaxRetries -> Exponent -> RetryPolicy
+exponentialBackoff maxRetries exponent =
+    limited maxRetries $ \retries currentSlot ->
+        Just $ mapSlot (\s -> s + floor (exponent ^^ (getSubmissionCount $ retries))) currentSlot
+
+-- | A very customisable resubmitter which can be configured with different
+-- retry policies.
+defaultResubmitFunction :: forall m. Monad m
+                        => (Core.TxAux -> m Bool)
+                        -> RetryPolicy
+                        -> ResubmissionFunction m
+defaultResubmitFunction send retryPolicy currentSlot scheduled oldSchedule = do
+    foldlM updateFn (emptyPending, oldSchedule) scheduled
+    where
+        updateFn :: (Evicted, Schedule) -> ScheduledTx -> m (Evicted, Schedule)
+        updateFn (evicted, acc@(Schedule s nursery)) entry = do
+            let (ScheduledTx txId txAux retries) = entry
+                reschedule targetSlot =
+                    IntMap.insertWith mappend
+                                      (castSlot targetSlot)
+                                      [incSubmissionCount entry]
+                                      s
+            case retryPolicy retries currentSlot of
+                 Nothing -> return (singletonPending txId txAux `unionPending` evicted, acc)
+                 Just newSlot -> do
+                     result <- send txAux
+                     case result of
```

Comment:
Why this pattern match? If sending fails, do we really want to give up on it?

Reply:
> If sending fails, do we really want to give up on it?

We don't, I guess, and my intention with that pattern match was to capture that spirit:

```haskell
                     case result of
                         False -> return (evicted, Schedule (reschedule newSlot) nursery)
                         True  -> return (evicted, acc)

```

I thought it would have been a good choice to simply "drop" the transaction if submission succeeded or to re-add it to the scheduler (according to the retry policy) in case we fail sending. What did I miss?

Reply:
Ah, then it's the _other_ case I should have asked about. We should never drop transactions here, should we? If they succeeded in transmission that doesn't mean that they'll make it into the blockchain. It's the responsibility of the wallet to inform us when they do.